pallets/gear-debug/src/tests.rs

```diff
@@ -300,3 +282,227 @@ fn get_last_message_id() -> MessageId {
+                    ;; check page 9 data
+                    i32.const 0x98765
+                    i32.load
+                    i32.const 0x42
+                    i32.eq
+                    br_if 0
+                    unreachable
+                )
+                ;; change page 3 and 4
+                ;; because we store 0x00_00_00_42 then bits will be changed
+                ;; in 3th page only, so the 3th page only must be in storage.
+                i32.const 0x3fffd
+                i32.const 0x42
+                i32.store
+            )
+
+            (data $.rodata (i32.const 0x10000) "0123456789")
+            (data $.rodata (i32.const 0x70001) "GEAR TECH")
+        )
+    "#;
+
+    init_logger();
+    new_test_ext().execute_with(|| {
+        let code = parse_wat(wat);
+        let program_id = generate_program_id(&code);
+        let origin = Origin::signed(1);
+
+        PalletGear::<Test>::submit_program(
+            origin.clone(),
+            code.clone(),
+            b"salt".to_vec(),
+            Vec::new(),
+            100_000_000_u64,
+            0_u128,
+        )
+        .expect("Failed to submit program");
+
+        // Enable debug-mode
+        DebugMode::<Test>::put(true);
+
+        run_to_block(2, None);
+
+        Pallet::<Test>::do_snapshot();
+
+        let static_pages = WasmPageNumber(8);
+
+        let page1_addr = 0x10000;
+        let gear_page1 = PageNumber::new_from_addr(page1_addr);
+        let mut page1_data = PageBuf::new_zeroed();
+        page1_data[..10].copy_from_slice(b"0123456780".as_slice());
+
+        let page8_addr = 0x87654;
+        let gear_page8 = PageNumber::new_from_addr(page8_addr);
+        let mut page8_data = PageBuf::new_zeroed();
+        page8_data[page8_addr % PageNumber::size()] = 0x42;
+
+        let page9_addr = 0x98765;
+        let gear_page9 = PageNumber::new_from_addr(page9_addr);
+        let mut page9_data = PageBuf::new_zeroed();
+        page9_data[page9_addr % PageNumber::size()] = 0x42;
+
+        let mut persistent_pages = BTreeMap::new();
+        persistent_pages.insert(gear_page1, page1_data.to_vec());
+        persistent_pages.insert(gear_page8, page8_data.to_vec());
+        persistent_pages.insert(gear_page9, page9_data.to_vec());
+
+        System::assert_last_event(
+            crate::Event::DebugDataSnapshot(DebugData {
+                dispatch_queue: vec![],
+                programs: vec![crate::ProgramDetails {
+                    id: program_id,
+                    state: crate::ProgramState::Active(crate::ProgramInfo {
+                        static_pages,
+                        persistent_pages: persistent_pages.clone(),
+                        code_hash: generate_code_hash(&code),
+                    }),
+                }],
+            })
+            .into(),
+        );
+
+        PalletGear::<Test>::send_message(origin, program_id, vec![], 10_000_000_u64, 0_u128)
+            .expect("Failed to send message");
```

Comment:
it's better to call any extrinsic under macros:
`assert_ok!` for this case, and `assert_noop!`, `assert_err!` for others.

Reply:
What's the difference ?

Reply:
`assert_noop` checks that the storage didn't changed while extrinsic failed. Others are just more idiomatic to use in substrate tests.

Reply:
Ok change to assert_ok!